File: src/notify.py
"""Telegram notifier for stock alerts."""
from __future__ import annotations

import os

import httpx

API_BASE = "https://api.telegram.org"


def _require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"{name} environment variable is not set")
    return value


def send(text: str, *, product_url: str | None = None, silent: bool = False) -> None:
    token = _require_env("TELEGRAM_BOT_TOKEN")
    chat_id = _require_env("TELEGRAM_CHAT_ID")

    payload: dict = {
        "chat_id": chat_id,
        "text": text,
        "parse_mode": "HTML",
        "disable_web_page_preview": False,
        "disable_notification": silent,
    }

    if product_url:
        payload["reply_markup"] = {
            "inline_keyboard": [
                [{"text": "상품 페이지 열기", "url": product_url}],
            ]
        }

    url = f"{API_BASE}/bot{token}/sendMessage"
    with httpx.Client(timeout=15) as client:
        response = client.post(url, json=payload)
        response.raise_for_status()


if __name__ == "__main__":
    send(
        "✅ <b>테스트 알림</b>\n알림 채널이 정상 작동합니다.",
        product_url="https://www.fujifilm-korea.co.kr/products/id/1330",
        silent=True,
    )

File: cloudflare/src/notify.ts
export interface TelegramSecrets {
  TELEGRAM_BOT_TOKEN: string;
  TELEGRAM_CHAT_ID: string;
}

export interface SendOptions {
  productUrl?: string;
  silent?: boolean;
}

export async function sendTelegram(
  secrets: TelegramSecrets,
  text: string,
  options: SendOptions = {},
): Promise<void> {
  const url = `https://api.telegram.org/bot${secrets.TELEGRAM_BOT_TOKEN}/sendMessage`;
  const payload: Record<string, unknown> = {
    chat_id: secrets.TELEGRAM_CHAT_ID,
    text,
    parse_mode: "HTML",
    disable_web_page_preview: false,
    disable_notification: options.silent ?? false,
  };
  if (options.productUrl) {
    payload.reply_markup = {
      inline_keyboard: [[{ text: "상품 페이지 열기", url: options.productUrl }]],
    };
  }
  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    const body = await response.text();
    throw new Error(`telegram api ${response.status}: ${body}`);
  }
}
